numba/core/ir_utils.py

```diff
@@ -2256,3 +2266,43 @@ def convert_code_obj_to_function(code_obj, caller_ir):
     # create the function and return it
     return _create_function_from_code_obj(fcode, func_env, func_arg, func_clo,
                                           glbls)
+
+
+def fixup_var_define_in_scope(blocks):
+    """Fixes the mapping of ir.Block to ensure all ir.Var are defined in all
+    scope.
+
+    Note: This is a workaround.
+    """
+    # Scan for all used variables
+    used_var = {}
+    for blk in blocks.values():
+        scope = blk.scope
+        for inst in blk.body:
+            for var in inst.list_vars():
+                used_var[var] = inst
+    # Note: not all blocks share a single scope even though they should.
+    # Ensure the scope of each block defines all used variables.
```

Comment:
Should the inverse also be checked? i.e. that there aren't variables in the scope(s) that don't exist in _some_ block. I'm not sure how likely it is for this issue to occur if at all. Am also not sure if this will end up causing more problems if there are cases where blocks are not sharing the same scope.

Reply:
It probably won't manifest as a problem but it is sign that the passes are not cleaning up properly. I am reluctant to add more checks for now as it is going to be a week long rabbit hole.

Reply:
Agree, until all passes conform/something else helps out with this, it's probably not a good use of time trying to debug it.